[WIP][inductor] Respect masking on zero-size tensors

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107008



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

Differential Revision: [D48256373](https://our.internmc.facebook.com/intern/diff/D48256373)

davidberard98 committed:
[WIP][inductor] Respect masking on zero-size tensors

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor] Respect masking on zero-size tensors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48256373).

aakhundov commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1283,7 +1284,10 @@ def load(self, name: str, index: sympy.Expr):
+            # Triton has better performance if you can provide a scalar index
+            # instead of a broadcasted index tensor, but we can only do this
+            # when there is no masking required.
+            if isinstance(original_index, sympy.Integer) and not mask_vars:
```
Seems that this code didn't take empty inputs into account. However, I'm wondering if for non-empty inputs "unmasked" loading with `original_index` would make sense to keep? Or do you think that if there are `mask_vars` they must be always codegened to be on the safe side? I guess, the latter would be more robust / conservative.

davidberard98 replied:
I felt like we should keep the mask if mask_vars exist. It seems like loads with scalar indices typically have less masks (e.g. usually loads have a mask like `xmask`, but this mask is not added for scalar indices)

peterbell10 replied:
I'm not sure this change really makes sense. The mask is there to prevent out of bounds memory accesses, which isn't an issue here. It also means that the loaded data can't be treated as a scalar by the triton compiler which may negatively impact perf.

davidberard98 replied:
@peterbell10 Out of bound memory access is the motivation for this PR - just added a test demonstrating the issue. If you run the test without the changes in this PR, an IMA is encountered

davidberard98 committed:
Update on "[WIP][inductor] Respect masking on zero-size tensors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48256373](https://our.internmc.facebook.com/intern/diff/D48256373)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor] Respect masking on zero-size tensors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48256373](https://our.internmc.facebook.com/intern/diff/D48256373)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48256373).

peterbell10 commented on test/inductor/test_custom_lowering.py:
```diff
@@ -0,0 +1,140 @@
+                        ops.index_expr(jagged_idx, offsets_dtype),
+                        end_idx,
+                    ),
+                    lambda: inp_loader([jagged_idx, emb_idx]),
```
Feels to me like the loader for a zero-size tensor should just return a constant NaN or something. Emitting a load instruction that will always be masked out doesn't make sense to me.

davidberard98 replied:
@peterbell10 after thinking about your comment for a bit, I still feel like this is a reasonable fix - I put more details/discussion in #107066 if you have any other comments.

I agree that yes, ideally we should convert this load to a constant/tl.full/etc. However, I think it would be a much more involved change; and I suspect the perf difference will be pretty minimal (but I'll double check perf once I get tests passing)

peterbell10 replied:
The change seems relatively simple to me, see #107074.

davidberard98 committed:
Update on "[WIP][inductor] Respect masking on zero-size tensors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48256373](https://our.internmc.facebook.com/intern/diff/D48256373)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor] Respect masking on zero-size tensors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48256373](https://our.internmc.facebook.com/intern/diff/D48256373)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor] Respect masking on zero-size tensors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48256373](https://our.internmc.facebook.com/intern/diff/D48256373)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor] Respect masking on zero-size tensors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48256373](https://our.internmc.facebook.com/intern/diff/D48256373)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor] Respect masking on zero-size tensors"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48256373](https://our.internmc.facebook.com/intern/diff/D48256373)

[ghstack-poisoned]

aakhundov commented:
@davidberard98 Wondering why is this PR closed? Is this intended?

peterbell10 commented:
The closing commit fixes the issue in a slightly different way, making this PR unnecessary.
